src/main/java/org/spongepowered/api/tag/TagTemplate.java

```diff
@@ -0,0 +1,108 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.tag;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.spongepowered.api.ResourceKey;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.datapack.DataPackSerializable;
+
+import java.util.Collection;
+
+/**
+ * A template that creates a or modifies a {@link Tag}.
+ *
+ * <p>All registrations through the API will generate into the Vanilla data pack system</p>
+ */
+public interface TagTemplate extends DataPackSerializable {
+
+    /**
+     * Creates a builder to create a new {@link Tag}
+     * or modify an existing one.
+     * @return Builder to build a new Tag.
+     */
+    @SuppressWarnings("unchecked")
+    static Builder<?> builder() {
+        return Sponge.game().builderProvider().provide(Builder.class);
+    }
+
+    interface Builder<T extends Taggable<T>> extends org.spongepowered.api.util.ResourceKeyedBuilder<TagTemplate, Builder<T>> {
+
+        /**
+         * Sets the tag type and generifies the builder.
+         *
+         * @param tagType Type to build a tag for, e.g {@link TagTypes#BLOCK_TYPE}
+         * @return This builder, for chaining
+         */
+        <NT extends Taggable<NT>> Builder<NT> type(TagType<NT> tagType);
+
+        /**
+         * <p>Whether to replace instead of append if the tag already
+         * exists. This replaces any minecraft tag data and any data packs
+         * that are lower priority than this one (the plugin data pack)</p>
+         *
+         * <p>By default, if the {@link ResourceKey#value()} is the same as
+         * another data pack / plugin, the two will be combined, and their
+         * tags appended to each other.</p>
+         *
+         * @param replace Whether to replace instead of merge.
+         * @return This builder, for chaining.
+         */
+        Builder<T> replace(boolean replace);
+
+        default Builder<T> replace() {
+            return replace(true);
+        }
+
+        /**
+         * Adds a child tag to this tag.
+         * This adds all of its values to this tag.
+         *
+         * @param childTag Child tag to add.
+         * @return This builder, for chaining
+         */
+        Builder<T> addChild(Tag<T> childTag);
```

Comment:
This will need be a template, too early in the bootstrap otherwise.
In fact, can you take a RegistryReference<TagTemplate>? Notice how in `WorldTemplate` I do not point at a `WorldType` directly.

Reply:
Correct. Infact it needs to be a RegistryKey to allow types that are registered by plugins/datapacks other than itself. Im not using a RegistryReference because what we actually need is a ResourceKey, guaranteed to be from the correct registry - we don't actually want the Tag object itself.